src/context/defrag.rs

```diff
@@ -0,0 +1,226 @@
+    /// The function is exposed for users that wants to implement the defrag function
+    /// on their module datatype, they can use this function to create [`DefragContext`]
+    /// that can be used in a safely manner.
+    /// Notice that the returned [`DefragContext`] borrows the pointer to [`raw::RedisModuleDefragCtx`]
+    /// so it can not outlive it (this means that it should not be used once the defrag callback ends).
+    pub unsafe fn new(defrag_ctx: *mut raw::RedisModuleDefragCtx) -> DefragContext {
+        DefragContext { defrag_ctx }
+    }
+
+    /// When the data type defrag callback iterates complex structures, this
+    /// function should be called periodically. A [`false`] return
+    /// indicates the callback may continue its work. A [`true`]
+    /// indicates it should stop.
+    ///
+    /// When stopped, the callback may use [`Self::set_cursor`] to store its
+    /// position so it can later use [`Self::get_cursor`] to resume defragging.
+    ///
+    /// When stopped and more work is left to be done, the callback should
+    /// return 1. Otherwise, it should return 0.
+    ///
+    /// NOTE: Modules should consider the frequency in which this function is called,
+    /// so it generally makes sense to do small batches of work in between calls.
+    pub fn should_stop(&self) -> bool {
+        let should_stop = unsafe {
+            RedisModule_DefragShouldStop.expect("RedisModule_DefragShouldStop is NULL")(
+                self.defrag_ctx,
+            )
+        };
+        should_stop != 0
+    }
+
+    /// Store an arbitrary cursor value for future re-use.
+    ///
+    /// This should only be called if [`Self::should_stop`] has returned a non-zero
+    /// value and the defrag callback is about to exit without fully iterating its
+    /// data type.
+    ///
+    /// This behavior is reserved to cases where late defrag is performed. Late
+    /// defrag is selected for keys that implement the `free_effort` callback and
+    /// return a `free_effort` value that is larger than the defrag
+    /// 'active-defrag-max-scan-fields' configuration directive.
+    ///
+    /// Smaller keys, keys that do not implement `free_effort` or the global
+    /// defrag callback are not called in late-defrag mode. In those cases, a
+    /// call to this function will return [`Status::Err`].
+    ///
+    /// The cursor may be used by the module to represent some progress into the
+    /// module's data type. Modules may also store additional cursor-related
+    /// information locally and use the cursor as a flag that indicates when
+    /// traversal of a new key begins. This is possible because the API makes
+    /// a guarantee that concurrent defragmentation of multiple keys will
+    /// not be performed.
+    pub fn set_cursor(&self, cursor: u64) -> Status {
+        unsafe {
+            RedisModule_DefragCursorSet.expect("RedisModule_DefragCursorSet is NULL")(
```

Comment:
`RedisModule_DefragCursorSet.expect("RedisModule_DefragCursorSet is available")`